src/clips-specs/rcll-central/production-strategy.clp

```diff
@@ -1372,26 +1417,177 @@
+  ; adjust existing goals that already received a priority update
+  (bind ?old-prio-val 0)
+  (do-for-fact ((?old-prio wm-fact)) (eq ?old-prio:key (create$ strategy meta priority increase buffer-cs args? mps ?cs))
+    (bind ?old-prio-val ?old-prio:value)
+    (retract ?old-prio)
+  )
+  (assert (wm-fact (key strategy meta priority increase buffer-cs args? mps ?cs) (value ?priority)))
+  (bind ?priority (- ?priority ?old-prio-val))
+  (delayed-do-for-all-facts ((?goal goal))
+    (and
+      (eq ?goal:mode FORMULATED)
+      (eq ?goal:class BUFFER-CAP)
+      (> ?goal:priority ?old-prio-val)
+      (member$ ?cs ?goal:params)
+      (member$ ?cap-color ?goal:params)
+    )
+    (modify ?goal (priority (+ ?goal:priority ?priority)))
+  )
+)
+
+(defrule production-strategy-revert-priority-to-buffer-cs
+  ?wf <- (wm-fact (key strategy meta priority increase buffer-cs args? mps ?cs) (value ?priority))
+  (not (wm-fact (key wp meta next-machine args? wp ?wp) (value ?cs)))
+  =>
+  (delayed-do-for-all-facts ((?goal goal))
+    (and
+      (eq ?goal:mode FORMULATED)
+      (eq ?goal:class BUFFER-CAP)
+      (> ?goal:priority ?priority)
+      (member$ ?cs ?goal:params)
+    )
+    (modify ?goal (priority (- ?goal:priority ?priority)))
+  )
+  (retract ?wf)
+)
+
+(defrule production-strategy-apply-priority-to-buffer-cs
+  (wm-fact (key strategy meta priority increase buffer-cs args? mps ?mps) (value ?prio-increase))
+  ?g <- (goal (class BUFFER-CS)
+    (mode FORMULATED) (priority ?prio) (params $? target-mps ?mps $?))
+  (test (< ?prio ?prio-increase))
+  =>
+  (modify ?g (priority (+ ?prio ?prio-increase)))
+)
+
+; -- increase priority to pay for a workpiece that waits
+
+(defrule production-strategy-increase-priority-to-pay-for-wp
+  "If a WP needs payment for it's next step, bump all respective payment goals"
+  (domain-fact (name mps-type) (param-values ?mps RS))
+  (wm-fact (key order meta wp-for-order args? wp ?wp ord ?order))
+  (domain-fact (name order-complexity) (param-values ?ord ?com))
+  (wm-fact (key wp meta next-step args? wp ?wp)
+    (value ?ring&:(str-index RING ?ring)))
+  (domain-fact (name ?order-ring-color) (param-values ?order ?ring-color))
+  (test
+    (eq ?order-ring-color (sym-cat order-ring (sub-string 5 5 ?ring) -color))
+  )
+  (domain-fact (name rs-ring-spec) (param-values ?mps ?ring-color ?ring-spec))
+  (domain-fact (name rs-filled-with) (param-values ?mps ?filled-with))
+  (test (< (sym-to-int ?filled-with) (sym-to-int ?ring-spec)))
+
+  (not (wm-fact (key strategy meta priority increase pay-for-wp args? wp ? mps ?mps $) (value ?old-prio&:(> ?old-prio (dynamic-prio-from-complexity ?com ?ring)))))
+  =>
+  (bind ?priority (dynamic-prio-from-complexity ?com ?ring))
+  ; adjust existing goals that already received a priority update
+  (bind ?old-prio-val 0)
+  (do-for-fact ((?old-prio wm-fact)) (and (wm-key-prefix ?old-prio:key (create$ strategy meta priority increase pay-for-wp))
```

Comment:
little bit confusing as you are overwriting variables here

Reply:
the variable is out of scope at this point in time, but you have a point. To increase readability I renamed the one in precondition to higher-prio